compile grails-shell, grails-gradle-model and grails-bootstrap w/ Gradle Groovy version

jamesfredley opened the pull request:
compile `grails-shell`, `grails-gradle-model` and `grails-bootstrap` with the Groovy version provided by Gradle to ensure build compatibility with Gradle, currently Groovy 3.0.x see: https://docs.gradle.org/current/userguide/compatibility.html#groovy

This will prevent running into Groovy 4 vs 3 compatibility issues like: https://github.com/bertramdev/asset-pipeline/pull/350, when Gradle executes the build using it's embedded Groovy Version, currently 3.0.22.

Matches changes on https://github.com/grails/grails-gradle-plugin/pull/334

jamesfredley committed:
compile grails-shell and grails-gradle-model with the Groovy version provided by Gradle

matrei reviewed (commented):
> grails-bootstrap was not changed due to Groovy XML package changes in 4.

Aren't the XML classes available in both `groovy.util` and `groovy.xml` in Groovy 3.
Are there other package changes that are incompatible?

jamesfredley committed:
Update grails-boostrap

jamesfredley commented:
PR updated to include grails-bootstap.  I wasn't aware of the dual packaging for XML and must have been missing a dependency during the first attempt.

matrei commented on grails-shell/build.gradle:
```diff
@@ -20,7 +27,7 @@ dependencies {
     api "org.gradle:gradle-tooling-api:$gradleToolingApiVersion"
-    compileOnly "org.apache.groovy:groovy-templates:$groovyVersion"
+
     api "org.apache.groovy:groovy-xml:$groovyVersion"
```
How about this one?

jamesfredley replied:
Yes, I left this one as org.apache.groovy and 4.0.x since it was not for compilation and is needed by at least one project down the line that is running it with Groovy 4 (not grails-gradle-plugin)

We did the same on https://github.com/bertramdev/asset-pipeline/blob/5.0.x/asset-pipeline-core/build.gradle#L56-L74

matrei replied:
Ok, but the one in `asset-pipeline` is `testImplementation`. This one is `api`. Doesn't `api` put it on the `compileClasspath`?

matrei reviewed (commented):
Nice! However, there is another dependency on `org.apache.groovy:groovy-xml`. Is that supposed to be left in?

matrei commented:
Don't we also have to exclude these projects from the Groovy 3 -> Groovy 4 substitution in the root `build.gradle` or they will just be replaced with Groovy 4 again?

jamesfredley committed:
Merge branch '7.0.x' into jamesfredley/fix-use-gradle-groovy-version

jamesfredley committed:
Remove code that forces Groovy 4 now that dependencies have been updated

jamesfredley committed:
cleanup

jamesfredley committed:
cleanup

jamesfredley committed:
restore groovy 4 configuration

jamesfredley committed:
Adjust dependency scope

jamesfredley committed:
more version cleanup

jamesfredley committed:
more cleanup

jamesfredley committed:
cleanup

jamesfredley committed:
update github workflow

jamesfredley committed:
update github workflow

jamesfredley committed:
Remove code that forces Groovy 4 now that dependencies have been updated

jamesfredley committed:
try  --rerun-tasks

jamesfredley committed:
test with groovy 3 on test

jamesfredley committed:
force groovy 3 in grails-bootstrap

jamesfredley committed:
groovy 4.0.23

jamesfredley committed:
experiment

jamesfredley committed:
followup experiment

jamesfredley committed:
Revert "force groovy 3 in grails-bootstrap"

This reverts commit d1c094929a58b7ae04fea19ce9c1333eb5f255e8.

jamesfredley committed:
Update dependencies

jamesfredley committed:
Revert "try  --rerun-tasks"

This reverts commit 0d111537eb38219476dfa08b8181bf0944df3470.

jamesfredley committed:
Revert "Remove code that forces Groovy 4 now that dependencies have been updated"

This reverts commit 0051fff532d204ef532f9d9c72ba824cfe9904c3.

jamesfredley committed:
Revert "update github workflow"

This reverts commit 27af60b48351fe5431432a1511b55ebb15fc1238.

jamesfredley committed:
Revert "update github workflow"

This reverts commit 00731102c0057486be74b73f82e169c6ea46cedf.

matrei commented:
Related https://github.com/grails/grails-gradle-plugin/pull/337

matrei commented on build.gradle:
```diff
@@ -213,17 +213,19 @@ if (isReleaseVersion) {
-    // FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
+// FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
+// except in projects that will be run by Gradle during the build
+configure(allprojects - project(':grails-shell') - project(':grails-bootstrap') - project(':grails-gradle-model')){
```
☝️ This does not seem to take effect:
```console
./gradlew grails-bootstrap:dI --dependency org.apache.groovy:groovy-xml --configuration compileClasspath

> Task :grails-bootstrap:dependencyInsight
org.apache.groovy:groovy-xml:4.0.22
  Variant compile:
    | Attribute Name                 | Provided | Requested         |
    |--------------------------------|----------|-------------------|
    | org.gradle.status              | release  |                   |
    | org.gradle.category            | library  | library           |
    | org.gradle.libraryelements     | jar      | classes+resources |
    | org.gradle.usage               | java-api | java-api          |
    | org.gradle.dependency.bundling |          | external          |
    | org.gradle.jvm.environment     |          | standard-jvm      |
    | org.gradle.jvm.version         |          | 17                |
   Selection reasons:
      - Selected by rule
      - By constraint

org.apache.groovy:groovy-xml:4.0.22
+--- org.apache.groovy:groovy-bom:4.0.22
|    \--- org.apache.groovy:groovy-ant:4.0.22
|         +--- compileClasspath (requested org.codehaus.groovy:groovy-ant:3.0.22)
|         \--- org.apache.groovy:groovy-bom:4.0.22 (*)
\--- org.apache.groovy:groovy-templates:4.0.22
     +--- compileClasspath (requested org.codehaus.groovy:groovy-templates:3.0.22)
     \--- org.apache.groovy:groovy-bom:4.0.22 (*)

org.codehaus.groovy:groovy-xml:3.0.22 -> org.apache.groovy:groovy-xml:4.0.22
\--- compileClasspath

(*) - Indicates repeated occurrences of a transitive dependency subtree. Gradle expands transitive dependency subtrees only once per project; repeat occurrences only display the root of the subtree, followed by this annotation.
```
However, if you add it back in the `allprojects` block with the following condition:
```groovy
allprojects {

    // FORCE UPGRADE OF GROOVY IN DEPENDENCIES TO GROOVY 4
    // except in projects that will be run by Gradle during the build
    if (!(project.name in ['grails-bootstrap', 'grails-gradle-model', 'grails-shell'])) {
        configurations.configureEach {
            resolutionStrategy.eachDependency { DependencyResolveDetails details ->
                if (details.requested.group == 'org.codehaus.groovy' && details.requested.name != 'groovy-bom') {
                    details.useTarget(group: 'org.apache.groovy', name: details.requested.name, version: groovyVersion)
                }
            }
        }
    }
}
```
```console
./gradlew grails-bootstrap:dI --dependency org.apache.groovy:groovy-xml --configuration compileClasspath
Starting a Gradle Daemon, 1 incompatible Daemon could not be reused, use --status for details

> Task :grails-bootstrap:dependencyInsight
No dependencies matching given input were found in configuration ':grails-bootstrap:compileClasspath'

./gradlew grails-bootstrap:dI --dependency org.codehaus.groovy:groovy-xml --configuration compileClasspath

> Task :grails-bootstrap:dependencyInsight
org.codehaus.groovy:groovy-xml:3.0.22
  Variant compile:
    | Attribute Name                 | Provided | Requested         |
    |--------------------------------|----------|-------------------|
    | org.gradle.status              | release  |                   |
    | org.gradle.category            | library  | library           |
    | org.gradle.libraryelements     | jar      | classes+resources |
    | org.gradle.usage               | java-api | java-api          |
    | org.gradle.dependency.bundling |          | external          |
    | org.gradle.jvm.environment     |          | standard-jvm      |
    | org.gradle.jvm.version         |          | 17                |

org.codehaus.groovy:groovy-xml:3.0.22
\--- compileClasspath
```

jamesfredley replied:
Gradle can be a mystery some days.  I had something similar initially, but then wanted to transition it to more of a "gradle" way `configure(allprojects - project(':grails-shell') - project(':grails-bootstrap') - project(':grails-gradle-model'))` and did not realize it stopped actually adjusting dependencies properly.  I've switched to this implementation with the check in a method, since it is reused.

matrei commented on grails-bootstrap/build.gradle:
```diff
@@ -1,23 +1,29 @@
+    }
 
+    testImplementation "org.apache.groovy:groovy-xml:$groovyVersion"
+    testImplementation "org.apache.groovy:groovy-templates:$groovyVersion "
```
There is an extra space at the end.

jamesfredley committed:
PR feedback

jamesfredley committed:
fix groovydoc

jamesfredley commented:
Dependency Insight now looks good for the 3 subprojects.

```
.gradlew grails-gradle-model:dI --dependency groovy --configuration compileClasspath
.gradlew grails-shell:dI --dependency groovy --configuration compileClasspath
.gradlew grails-bootstrap:dI --dependency groovy --configuration compileClasspath
```

groovydocs required several changes, including moving where and how we set `JavaLanguageVersion.of(17)`, there are still some errors when JavaParser runs related to pattern and instanceOf.  These existed prior to these changes and do not stop the build.

I've tested `grails-core` -> `grails-gradle-plugin` -> `grails-profile base` to ensure that dependencies are still passed as expected.  `grails-profile base` is a good test case because it receives almost all of its dependencies from `grails-gradle-plugin`, which inherits some from `grails-boostrap` and `grails-shell` in `grails-core`.

jamesfredley commented:
grails-core does not use grails-gradle-plugin during build so https://github.com/grails/grails-gradle-plugin/pull/337 does not impact these changes.

matrei committed:
fix: get `groovydoc` task working again

Reverts some changes that updated to Java 17 syntax `instanceof`. At the moment this breaks groovydoc generation with the message:
Use of patterns with instanceof is not supported. Pay attention that this feature is supported starting from 'JAVA_14' language level. If you need that feature the language level must be configured in the configuration before parsing the source files.

As these changes are purely cosmetic I'm find it better to revert them.
